x/gov/simulation/msgs.go

```diff
@@ -18,30 +19,89 @@ const (
+		// 2.2) select who votes and when
+		whoVotes := r.Perm(len(keys))
+		// didntVote := whoVotes[numVotes:]
+		whoVotes = whoVotes[:numVotes]
+		votingPeriod := k.GetVotingProcedure(ctx).VotingPeriod
+		fops := make([]simulation.FutureOperation, numVotes+1)
+		for i := 0; i < numVotes; i++ {
+			whenVote := ctx.BlockHeight() + r.Int63n(votingPeriod)
+			fops[i] = simulation.FutureOperation{int(whenVote), operationSimulateMsgVote(k, sk, keys[whoVotes[i]], proposalID)}
+		}
+		// 3) Make an operation to ensure slashes were done correctly. (Really should be a future invariant)
+		// TODO: Find a way to check if a validator was slashed other than just checking their balance a block
+		// before and after.
+
+		return action, fops, nil
+	}
+}
+
 // SimulateMsgSubmitProposal simulates a msg Submit Proposal
 // Note: Currently doesn't ensure that the proposal txt is in JSON form
 func SimulateMsgSubmitProposal(k gov.Keeper, sk stake.Keeper) simulation.Operation {
 	handler := gov.NewHandler(k)
 	return func(tb testing.TB, r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, keys []crypto.PrivKey, log string, event func(string)) (action string, fOps []simulation.FutureOperation, err sdk.Error) {
-		msg := simulationCreateMsgSubmitProposal(tb, r, keys, log)
-		ctx, write := ctx.CacheContext()
-		result := handler(ctx, msg)
-		if result.IsOK() {
-			// Update pool to keep invariants
-			pool := sk.GetPool(ctx)
-			pool.LooseTokens = pool.LooseTokens.Sub(sdk.NewDecFromInt(msg.InitialDeposit.AmountOf(denom)))
-			sk.SetPool(ctx, pool)
-			write()
-		}
-		event(fmt.Sprintf("gov/MsgSubmitProposal/%v", result.IsOK()))
-		action = fmt.Sprintf("TestMsgSubmitProposal: ok %v, msg %s", result.IsOK(), msg.GetSignBytes())
+		sender := simulation.RandomKey(r, keys)
+		msg := simulationCreateMsgSubmitProposal(tb, r, sender, log)
+		action = simulateHandleMsgSubmitProposal(msg, sk, handler, ctx, event)
 		return action, nil, nil
 	}
 }
 
-func simulationCreateMsgSubmitProposal(tb testing.TB, r *rand.Rand, keys []crypto.PrivKey, log string) gov.MsgSubmitProposal {
-	key := simulation.RandomKey(r, keys)
-	addr := sdk.AccAddress(key.PubKey().Address())
+func simulateHandleMsgSubmitProposal(msg gov.MsgSubmitProposal, sk stake.Keeper, handler sdk.Handler, ctx sdk.Context, event func(string)) (action string) {
+	ctx, write := ctx.CacheContext()
+	result := handler(ctx, msg)
+	if result.IsOK() {
+		// Update pool to keep invariants
+		pool := sk.GetPool(ctx)
+		pool.LooseTokens = pool.LooseTokens.Sub(sdk.NewDecFromInt(msg.InitialDeposit.AmountOf(denom)))
+		sk.SetPool(ctx, pool)
+		write()
+	}
+	event(fmt.Sprintf("gov/MsgSubmitProposal/%v", result.IsOK()))
+	action = fmt.Sprintf("TestMsgSubmitProposal: ok %v, msg %s", result.IsOK(), msg.GetSignBytes())
+	return action
+}
+
+func simulationCreateMsgSubmitProposal(tb testing.TB, r *rand.Rand, sender crypto.PrivKey, log string) gov.MsgSubmitProposal {
+	addr := sdk.AccAddress(sender.PubKey().Address())
```

Comment:
In a future PR, we should make a simulation key object which has the privkey, pubkey, and address. The simulator spends ~3% of its time computing pubkeys. Improving this will allow us to simulate greater block sizes. (Once we reduce how often governance proposals are. The amount of slashes they induce is what causes the simulator to slow down at blocks > 200)

Reply:
Good idea